### PR TITLE
Add unary visitors, and most other scalar types to ffi expressions

### PR DESCRIPTION
**Pull request by nicklan**
As a follow-up to #231:

Add a `visit_unary` and make a macro for the other literal scalar types. Also adds some missing `From`s to `Scalar`.

Punting on date/timestamp/decimal for now, need to think more on those how we want to pass them.

### CODE REVIEW

File: ffi/src/lib.rs [hunk 1]
```diff
@@ -961,6 +962,19 @@ fn visit_expression_binary(
     }
 }
 
+fn visit_expression_unary(
+    state: &mut KernelExpressionVisitorState,
+    op: UnaryOperator,
+    inner_expr: usize,
+) -> usize {
+    if let Some(expr) = unwrap_kernel_expression(state, inner_expr).map(Box::new) {
+        wrap_expression(state, Expression::UnaryOperation { op, expr })
```
Comment (scovich) [inline ffi/src/lib.rs:971]: Seems like we should define `Expression::unary` that takes `impl Into<Expression>`, similar to the existing `Expression::binary`? Then this whole method simplifies to:
```rust
unwrap_kernel_expression(state, inner_expr)
    .map_or_else(0, |expr| wrap_expression(state, Expression::unary(op, expr))
```

**Review by hntd187 — APPROVED** [review]
LGTM

File: ffi/src/lib.rs [hunk 2]
```diff
@@ -1057,10 +1087,26 @@ fn visit_expression_literal_string_impl(
...
-    wrap_expression(state, Expression::Literal(Scalar::from(value)))
+macro_rules! fn_visit_literal_prim {
+    ( $(($name: ident, $typ: ty)), * ) => {
+        $(
+            #[no_mangle]
+            pub extern "C" fn $name(
+                state: &mut KernelExpressionVisitorState,
+                value: $typ,
+            ) -> usize {
+                wrap_expression(state, Expression::Literal(Scalar::from(value)))
```
Comment (scovich) [inline ffi/src/lib.rs:1098]: Why not just:
```suggestion
                wrap_expression(state, Expression::literal(value))
```

**Review by scovich — APPROVED** [review]
LGTM, couple nits

File: ffi/src/lib.rs (hunk 1 above)
Comment (nicklan) [reply ffi/src/lib.rs]: Yep, that's nicer. We actually already had `Expression::unary` :)

**Comment by samansmink** [comment]
I'm not getting the `visit_expression_literal_long` etc. functions exposed in the ffi as I would expect when building this from within duckdb delta. Maybe I'm missing something? i would expect all the `visit_expression_literal_*` functions to be there

**Comment by nicklan** [comment]
> I'm not getting the `visit_expression_literal_long` etc. functions exposed in the ffi as I would expect when building this from within duckdb delta. Maybe I'm missing something? i would expect all the `visit_expression_literal_*` functions to be there

Right you are. `cbindgen` doesn't pick them up because they are macros. Yar. I'll see if there's a way to make that work, or I'll unmacro them. sorry about that!

**Comment by nicklan** [comment]
> I'm not getting the `visit_expression_literal_long` etc. functions exposed in the ffi as I would expect when building this from within duckdb delta. Maybe I'm missing something? i would expect all the `visit_expression_literal_*` functions to be there

Fixed by #256

**Pull request merged**